### PR TITLE
Fix: inconsistent encoding in the REST controller #2687  and Uploading Entity with self-reference (xref) fails #2637 Fix Landing on a datae-explorer plugin after login results in errors #2495 [0.25]

### PR DESCRIPTION
**Pull request by jjettenn**
Fixing encoding.
Encoding was not consistent for the href attribute.
The encoding util for encoding the href attribute is changed because it
was wrong. Now using UriUtils of spring instead of the previus
URLEncoder that was using application/x-www-form-urlencoded instead of
RFC 3986

### CODE REVIEW

File: molgenis-data-mysql/src/main/java/org/molgenis/data/importer/ImportWriter.java [hunk 1]
```diff
@@ -507,4 +510,45 @@ public int update(Repository repo, Iterable<? extends Entity> entities, Database
...
+			super(entityMetaData, dataService, entity);
+		}
+
+		/**
+		 * getEntity returns null when attributeName is not resulting in an entity
+		 */
+		@Override
+		public Entity getEntity(String attributeName)
+		{
+			try {
```
Comment (fdlk) [inline molgenis-data-mysql/src/main/java/org/molgenis/data/importer/ImportWriter.java]: formatting :)

File: molgenis-data-mysql/src/main/java/org/molgenis/data/mysql/MysqlRepository.java [hunk 2]
```diff
@@ -949,7 +989,10 @@ public void deleteById(Iterable<Object> ids)
 	{
 		final List<Object> idList = new ArrayList<Object>();
 		for (Object id : ids)
+		{
 			idList.add(id);
+			System.out.println("MysqlRepositoy 956 id: " + id.toString());
```
Comment (fdlk) [inline molgenis-data-mysql/src/main/java/org/molgenis/data/mysql/MysqlRepository.java]: sysout

File: molgenis-data-mysql/src/main/java/org/molgenis/data/mysql/MysqlRepository.java [hunk 3]
```diff
@@ -923,10 +924,49 @@ public void delete(Entity entity)
...
+		for (Entity e : entities)
+		{
+			deleteByIdBatch.add(e.getIdValue());
+		}
+		this.deleteById(deleteByIdBatch);
+	}
+
+	/**
+	 * Use before a delete action of a entity with XREF data type were the entity and refEntity are the same entities.
+	 *
```
Comment (fdlk) [inline molgenis-data-mysql/src/main/java/org/molgenis/data/mysql/MysqlRepository.java]: were -> where

**Comment by fdlk** [comment]
I fixed and then tried to upload the typetest dataset. It fails stating that a foreign key constraint is not met. Could it be that cross-entity references are now broken?

**Comment by jjettenn** [comment]
The xcategorymrefnullable name was te groot
Op 25 mrt. 2015 23:10 schreef "Fleur Kelpin" <notifications@github.com>:

**Comment by fdlk** [comment]
Yes. So when I fixed that by making it smaller (see https://github.com/jjettenn/molgenis/pull/2) I got a foreign key constraint failure.

**Comment by Mark-de-Haan** [comment]
You need to drop the database, I think transactional import might be broken
cause I ran into this problem as well. Import fails -> fix EMX -> foreign
key constraints

On 26 March 2015 at 00:40, Fleur Kelpin <notifications@github.com> wrote:

> Yes. So when I fixed that by making it smaller (see jjettenn#2
> <https://github.com/jjettenn/molgenis/pull/2>) I got a foreign key
> constraint failure.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/molgenis/molgenis/pull/2689#issuecomment-86253916>.
>



-- 
Mark de Haan
Genomics Coordination Centre
Genetics Department
University Medical Centre Groningen

**Pull request merged**